implemented adding and deleting

olenitut opened the pull request:
[DEMO LINK](https://olenitut.github.io/react_todo-app-add-and-delete/)

witflash commented on src/components/Warning/warning.tsx:
```diff
@@ -0,0 +1,47 @@
+const Warning: React.FC<Props> = ({ message }) => {
+  const [isShown, setIsShown] = useState(false);
+
+  useEffect(() => {
+    if (message) {
+      setIsShown(true);
+      setTimeout(() => {
+        setIsShown(false);
+      }, 3000);
+    } else {
+      setIsShown(false);
+    }
+  }, [message]);
```
```suggestion
  useEffect(() => {
    let timer;
    
    if (message) {
      setIsShown(true);
      timer = setTimeout(() => {
        setIsShown(false);
      }, 3000);
    } else {
      setIsShown(false);
    }
    
    return () => {
      if (timer) {
        clearTimeout(timer);
      }
    }
  }, [message]);
```

You should clear any async timeouts in useEffect callback function.

witflash commented on src/components/Todo/Todo.tsx:
```diff
@@ -0,0 +1,70 @@
+  return (
+    <div
+      data-cy="Todo"
+      className={
+        classNames('todo', {
+          completed: todo.completed,
+        })
+      }
```
```suggestion
      className={classNames('todo', {
        completed: todo.completed,
      })}
```

So you will have more compact syntax

witflash commented on src/components/Todo/Todo.tsx:
```diff
@@ -0,0 +1,70 @@
+
+  const isActive = todo.completed && deletingCompleted ? true : hasSpinner;
+
+  useEffect(() => {
+    if (!isDeleting) {
+      setHasSpinner(false);
+    }
+  }, [isDeleting]);
```
You don't need this. You can check `isDeleting` directly.

witflash replied:
Because you handle this statement in `onDeleteTodo` already.

olenitut replied:
I implemented this commentary and the following ones, and that broke my code((( it seems that i do need this deleting completed and cannot do only with isDeleting, as deleted completed handles only the deleting of completed tasks and isDeleting with individual ones. Therefore when i set the spinner, i need to handle the two states separately. there may be different logic but i will stick to this one for now as i will break my brain trying to redo the logic of the whole app))

witflash commented on src/components/Todo/Todo.tsx:
```diff
@@ -0,0 +1,70 @@
+}) => {
+  const [hasSpinner, setHasSpinner] = useState(spinner);
+
+  const isActive = todo.completed && deletingCompleted ? true : hasSpinner;
```
```suggestion
  const isActive = todo.completed && !isDeleting;
```

Anastasiia-Svintsova replied:
Not fixed

witflash commented on src/components/Todo/Todo.tsx:
```diff
@@ -0,0 +1,70 @@
+  spinner: boolean;
+  onDeleteTodo: (todoId: number) => void;
+  isDeleting: boolean;
+  deletingCompleted: boolean;
```
You don't need this prop. You can use `isDeleting` only.

Anastasiia-Svintsova replied:
Not fixed

witflash commented on src/components/Todo/Todo.tsx:
```diff
@@ -0,0 +1,70 @@
+        type="button"
+        className="todo__remove"
+        data-cy="TodoDeleteButton"
+        onClick={() => deleteHandler(todo.id)}
```
And here you can do just:
```suggestion
        onClick={() => onDeleteTodo(todo.id)}
```

Anastasiia-Svintsova replied:
Not fixed

witflash reviewed (changes requested):
Please check my comments and add fixes.

olenitut commented:
HI! I am writing to the mentor who is gonna review this task. I already have the final version of the todo app in the other pull request : https://github.com/mate-academy/react_todo-app-with-api/pull/245 

if possible, please, approve this one and check the final version, as there I implemented many changes and also fixed some bugs present in this one. I dont want to rethink the logic here as I alsredy did it in the final version. Thanks!

Anastasiia-Svintsova commented on src/App.tsx:
```diff
@@ -1,18 +1,130 @@
+  const [todosFromTheServer, setTodosFromTheServer] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [todosToShow, setTodosToShow] = useState([...todosFromTheServer]);
+  const [currentFilter, setCurrentFilter] = useState('ALL');
```
Create an enum for filter types, please

Anastasiia-Svintsova commented on src/App.tsx:
```diff
@@ -1,18 +1,130 @@
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [todosFromTheServer, setTodosFromTheServer] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [todosToShow, setTodosToShow] = useState([...todosFromTheServer]);
```
Looks better for me, but it's up to you
```suggestion
  const [visibleTodos, setVisibleTodos] = useState([...todosFromTheServer]);
```

Anastasiia-Svintsova commented on src/App.tsx:
```diff
@@ -1,18 +1,130 @@
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+    filterTodos(currentFilter);
+  }, [todosFromTheServer]);
```
Do you really need `todosFromTheServer` in the dependency array? Looks like you need to remove it and add currentFilter and filterTodos instead. Additionally, wrap filterTodos in the useCallback

Anastasiia-Svintsova commented on src/App.tsx:
```diff
@@ -1,18 +1,130 @@
     }
-  }, []);
+
+    if (!user) {
+      return;
+    }
```
I think you'll have a user in any case 🤔 You can't show the form if a user is not login

Anastasiia-Svintsova commented on src/App.tsx:
```diff
@@ -24,200 +136,35 @@ export const App: React.FC = () => {
-          </button>
-        </footer>
+            <Footer
+              left={todosFromTheServer.filter(todo => !todo.completed).length}
+              completed={
+                todosFromTheServer.filter(todo => todo.completed).length
+              }
```
What is this? Move these values to variables, please. Or pass todosFromTheServer as props and count these values in the Footer directly

Anastasiia-Svintsova commented on src/api/todos.ts:
```diff
@@ -1,8 +1,22 @@
+  return client.delete(`/todos/${id}`);
+};
+
 // Add more methods here
```
Remove comments

Anastasiia-Svintsova commented on src/components/Footer/Footer.tsx:
```diff
@@ -0,0 +1,78 @@
+  left: number;
+  completed: number;
+  onFilter: (type: string) => void;
+  onClearCompleted: ()=>void;
```
```suggestion
  onClearCompleted: () => void;
```

Anastasiia-Svintsova commented on src/components/Footer/Footer.tsx:
```diff
@@ -0,0 +1,78 @@
+          className={classNames(
+            'filter__link', { selected: selectedFilter === 'ALL' },
+          )}
+          onClick={() => {
+            onFilter('ALL');
+            setSelectedFilter('ALL');
+          }}
```
Move to a function

Anastasiia-Svintsova commented on src/components/Footer/Footer.tsx:
```diff
@@ -0,0 +1,78 @@
+          className={classNames(
+            'filter__link', { selected: selectedFilter === 'ACTIVE' },
+          )}
+          onClick={() => {
+            onFilter('ACTIVE');
+            setSelectedFilter('ACTIVE');
+          }}
```
Same

Anastasiia-Svintsova commented on src/utils/fetchClient.ts:
```diff
@@ -1,6 +1,7 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Don't disable eslint, try to solve the error

Anastasiia-Svintsova reviewed (changes requested):
- Pay attention to my comments and fix the previous ones

olenitut commented:
HI! I am writing to the mentor who is gonna review this task. I already have the final version of the todo app in the other pull request : https://github.com/mate-academy/react_todo-app-with-api/pull/245

if possible, please, approve this one and check the final version, as there I implemented many changes and also fixed some bugs present in this one. I dont want to rethink the logic here as I alsredy did it in the final version. Thanks!
